### ISSUE TITLE
ERROR: [102]: unable to wait on child process: [10] No child processes

### ISSUE THREAD

**Issue by todeb** [pgbackrest/pgbackrest#1531]
Seems that I have similar issue to this https://github.com/pgbackrest/pgbackrest/issues/1205
Commented on that thread but as it is closed I'm openning new issue.

Running 2 patroni instances, and 1 remote repo host.

Backups are running ok, but suddenely crashing. On one or both patroni nodes, pgbackrest process hang and create 100% CPU load.

On the repo host I see  "No child processes" on first failing backup.
```
2021-10-06 01:00:03.862 P00   INFO: last backup label = 20211006-000004F, version = 2.35
2021-10-06 01:00:03.862 P00   INFO: execute non-exclusive pg_start_backup(): backup begins after the next regular checkpoint completes
2021-10-06 01:10:04.113 P00  ERROR: [102]: unable to wait on child process: [10] No child processes
```
On the next failed backups:
```
 unable to acquire lock on file '/tmp/pgbackrest/main-backup.lock': Resource temporarily unavailable
                                    HINT: is another pgBackRest process running?
```

DEBUG:
```
2021-10-06 02:33:47.655 P00  DEBUG:     common/exit::exitSafe: (result: 0, error: true, signalType: 0)
2021-10-06 02:33:47.656 P00  ERROR: [102]: unable to wait on child process: [10] No child processes
                                    --------------------------------------------------------------------
                                    If SUBMITTING AN ISSUE please provide the following information:

                                    version: 2.35
                                    command: backup
                                    options: --backup-standby --exec-id=2615-6a7b2a34 --log-level-file=debug --pg1-host=10.149.100.244 --pg2-host=10.149.100.245 --pg1-host-port=2222 --pg2-host-port=2222 --pg1-path=/opt/pgsql/patroni --pg2-path=/opt/pgsql/patroni --pg1-port=5432 --pg2-port=5432 --pg1-user=postgres --pg2-user=postgres --repo1-path=/opt/repo --repo1-retention-diff=30 --repo1-retention-full=2 --stanza=main --type=incr

                                    stack trace:
                                    common/exec.c:execCheck:175:(trace log level required for parameters)
                                    common/exec.c:execWrite:(trace log level required for parameters)
                                    common/io/write.c:ioWriteFlush:(trace log level required for parameters)
                                    protocol/client.c:protocolClientCommandPut:(trace log level required for parameters)
                                    protocol/client.c:protocolClientExecute:(trace log level required for parameters)
                                    db/db.c:dbFreeResource:(trace log level required for parameters)
                                    main.c:main:(debug log level required for parameters)
                                    --------------------------------------------------------------------
2021-10-06 02:33:47.656 P00   WARN: unable to wait on child process: [10] No child processes
2021-10-06 02:33:47.656 P00   WARN: unable to wait on child process: [10] No child processes
2021-10-06 02:33:47.656 P00   WARN: unable to wait on child process: [10] No child processes
2021-10-06 02:33:47.656 P00   INFO: backup command end: aborted with exception [102]
2021-10-06 02:33:47.656 P00  DEBUG:     common/lock::lockRelease: (failOnNoLock: false)
2021-10-06 02:33:47.656 P00  DEBUG:     common/lock::lockRelease: => true
```

TRACE:
```
2021-10-06 04:30:05.639 P00  TRACE:         common/exec::execCheck: (this: {Exec})
2021-10-06 04:30:05.646 P00  DEBUG:     common/exit::exitSafe: (result: 0, error: true, signalType: 0)
2021-10-06 04:30:05.647 P00  ERROR: [102]: unable to wait on child process: [10] No child processes
                                    --------------------------------------------------------------------
                                    If SUBMITTING AN ISSUE please provide the following information:

                                    version: 2.35
                                    command: backup
                                    options: --backup-standby --exec-id=3322-278ab4bd --log-level-file=trace --pg1-host=10.149.100.244 --pg2-host=10.149.100.245 --pg1-host-port=2222 --pg2-host-port=2222 --pg1-path=/opt/pgsql/patroni --pg2-path=/opt/pgsql/patroni --pg1-port=5432 --pg2-port=5432 --pg1-user=postgres --pg2-user=postgres --repo1-path=/opt/repo --repo1-retention-diff=30 --repo1-retention-full=2 --stanza=main --type=incr

                                    stack trace:
                                    common/exec.c:execCheck:175:(this: {Exec})
                                    common/exec.c:execWrite:(this: {Exec}, buffer: {used: 11, size: 1048576})
                                    common/io/write.c:ioWriteFlush:(this: {IoWrite})
                                    protocol/client.c:protocolClientCommandPut:(this: {name: remote-0 protocol on '10.149.100.245'}, command: {command: db-c})
                                    protocol/client.c:protocolClientExecute:(this: {name: remote-0 protocol on '10.149.100.245'}, command: {command: db-c}, resultRequired: false)
                                    db/db.c:dbFreeResource:(this: {client: null, remoteClient: {name: remote-0 protocol on '10.149.100.245'}})
                                    main.c:main:(debug log level required for parameters)
                                    --------------------------------------------------------------------
2021-10-06 04:30:05.649 P00  TRACE:         protocol/helper::protocolFree: (void)
2021-10-06 04:30:05.649 P00  TRACE:         protocol/helper::protocolRemoteFree: (hostIdx: 0)
2021-10-06 04:30:05.649 P00  TRACE:         protocol/helper::protocolHelperClientFree: (protocolHelperClient: *void)
2021-10-06 04:30:05.649 P00  TRACE:         protocol/client::protocolClientFreeResource: (this: {name: remote-0 protocol on '10.149.100.244'})
2021-10-06 04:30:05.649 P00  TRACE:         protocol/client::protocolClientCommandPut: (this: {name: remote-0 protocol on '10.149.100.244'}, command: {command: exit})
2021-10-06 04:30:05.649 P00  TRACE:         common/io/write::ioWrite: (this: {IoWrite}, buffer: {used: 8, size: 1048576})
2021-10-06 04:30:05.649 P00  TRACE:         common/io/filter/group::ioFilterGroupProcess: (this: {inputSame: false, done: false}, input: {used: 8, size: 1048576}, output: {used: 11, size: 1048576})
2021-10-06 04:30:05.649 P00  TRACE:         common/io/filter/buffer::ioBufferProcess: (this: {inputSame: false, inputPos: 0}, input: {used: 8, size: 1048576}, output: {used: 11, size: 1048576})
2021-10-06 04:30:05.649 P00  TRACE:         common/io/filter/buffer::ioBufferProcess: => void
2021-10-06 04:30:05.649 P00  TRACE:         common/io/filter/group::ioFilterGroupProcess: => void
2021-10-06 04:30:05.649 P00  TRACE:         common/io/write::ioWrite: => void
2021-10-06 04:30:05.649 P00  TRACE:         common/io/write::ioWriteFlush: (this: {IoWrite})
2021-10-06 04:30:05.649 P00  TRACE:         common/exec::execWrite: (this: {Exec}, buffer: {used: 19, size: 1048576})
2021-10-06 04:30:05.649 P00  TRACE:         common/io/write::ioWrite: (this: {IoWrite}, buffer: {used: 19, size: 1048576})
2021-10-06 04:30:05.649 P00  TRACE:         common/io/filter/group::ioFilterGroupProcess: (this: {inputSame: false, done: false}, input: {used: 19, size: 1048576}, output: {used: 11, size: 1048576})
2021-10-06 04:30:05.649 P00  TRACE:         common/io/filter/buffer::ioBufferProcess: (this: {inputSame: false, inputPos: 0}, input: {used: 19, size: 1048576}, output: {used: 11, size: 1048576})
2021-10-06 04:30:05.649 P00  TRACE:         common/io/filter/buffer::ioBufferProcess: => void
2021-10-06 04:30:05.649 P00  TRACE:         common/io/filter/group::ioFilterGroupProcess: => void
2021-10-06 04:30:05.649 P00  TRACE:         common/io/write::ioWrite: => void
2021-10-06 04:30:05.649 P00  TRACE:         common/io/write::ioWriteFlush: (this: {IoWrite})
2021-10-06 04:30:05.649 P00  TRACE:         common/io/fdWrite::ioFdWrite: (this: {IoFdWrite}, buffer: {used: 30, size: 1048576})
2021-10-06 04:30:05.649 P00  TRACE:         common/exec::execCheck: (this: {Exec})
2021-10-06 04:30:05.649 P00   WARN: unable to wait on child process: [10] No child processes
2021-10-06 04:30:05.649 P00  TRACE:         common/exec::execFreeResource: (this: {Exec})
2021-10-06 04:30:05.649 P00  TRACE:         common/exec::execFreeResource: => void
2021-10-06 04:30:05.649 P00  TRACE:         protocol/helper::protocolHelperClientFree: => void
2021-10-06 04:30:05.649 P00  TRACE:         protocol/helper::protocolRemoteFree: => void
2021-10-06 04:30:05.649 P00  TRACE:         protocol/helper::protocolRemoteFree: (hostIdx: 1)
2021-10-06 04:30:05.649 P00  TRACE:         protocol/helper::protocolHelperClientFree: (protocolHelperClient: *void)
2021-10-06 04:30:05.649 P00  TRACE:         protocol/client::protocolClientFreeResource: (this: {name: remote-0 protocol on '10.149.100.245'})
2021-10-06 04:30:05.649 P00  TRACE:         protocol/client::protocolClientCommandPut: (this: {name: remote-0 protocol on '10.149.100.245'}, command: {command: exit})
2021-10-06 04:30:05.649 P00  TRACE:         common/io/write::ioWrite: (this: {IoWrite}, buffer: {used: 8, size: 1048576})
2021-10-06 04:30:05.649 P00  TRACE:         common/io/filter/group::ioFilterGroupProcess: (this: {inputSame: false, done: false}, input: {used: 8, size: 1048576}, output: {used: 11, size: 1048576})
2021-10-06 04:30:05.649 P00  TRACE:         common/io/filter/buffer::ioBufferProcess: (this: {inputSame: false, inputPos: 0}, input: {used: 8, size: 1048576}, output: {used: 11, size: 1048576})
2021-10-06 04:30:05.649 P00  TRACE:         common/io/filter/buffer::ioBufferProcess: => void
2021-10-06 04:30:05.649 P00  TRACE:         common/io/filter/group::ioFilterGroupProcess: => void
2021-10-06 04:30:05.649 P00  TRACE:         common/io/write::ioWrite: => void
2021-10-06 04:30:05.649 P00  TRACE:         common/io/write::ioWriteFlush: (this: {IoWrite})
2021-10-06 04:30:05.649 P00  TRACE:         common/exec::execWrite: (this: {Exec}, buffer: {used: 19, size: 1048576})
2021-10-06 04:30:05.649 P00  TRACE:         common/io/write::ioWrite: (this: {IoWrite}, buffer: {used: 19, size: 1048576})
2021-10-06 04:30:05.649 P00  TRACE:         common/io/filter/group::ioFilterGroupProcess: (this: {inputSame: false, done: false}, input: {used: 19, size: 1048576}, output: {used: 11, size: 1048576})
2021-10-06 04:30:05.649 P00  TRACE:         common/io/filter/buffer::ioBufferProcess: (this: {inputSame: false, inputPos: 0}, input: {used: 19, size: 1048576}, output: {used: 11, size: 1048576})
2021-10-06 04:30:05.649 P00  TRACE:         common/io/filter/buffer::ioBufferProcess: => void
2021-10-06 04:30:05.649 P00  TRACE:         common/io/filter/group::ioFilterGroupProcess: => void
2021-10-06 04:30:05.649 P00  TRACE:         common/io/write::ioWrite: => void
2021-10-06 04:30:05.649 P00  TRACE:         common/io/write::ioWriteFlush: (this: {IoWrite})
2021-10-06 04:30:05.649 P00  TRACE:         common/io/fdWrite::ioFdWrite: (this: {IoFdWrite}, buffer: {used: 30, size: 1048576})
2021-10-06 04:30:05.649 P00  TRACE:         common/exec::execCheck: (this: {Exec})
2021-10-06 04:30:05.649 P00   WARN: unable to wait on child process: [10] No child processes
2021-10-06 04:30:05.649 P00  TRACE:         common/exec::execFreeResource: (this: {Exec})
2021-10-06 04:30:05.649 P00  TRACE:         common/wait::waitNew: (waitTime: 1830000)
2021-10-06 04:30:05.649 P00  TRACE:         common/wait::waitNew: => {Wait}
2021-10-06 04:30:05.649 P00   WARN: unable to wait on child process: [10] No child processes
```

What I see on postgres nodes when process hangs: 
![image](https://user-images.githubusercontent.com/61552048/136165295-87e37e26-b408-48f0-b698-43ddd42d19b4.png)

pgbackrest info about completed backups:
```
        incr backup: 20211006-000004F_20211006-025605I
            timestamp start/stop: 2021-10-06 02:56:05 / 2021-10-06 03:01:06
            wal start/stop: 00000085000000E10000003A / 00000085000000E10000003A
            database size: 10.2GB, database backup size: 1.8GB
            repo1: backup set size: 1.8GB, backup size: 172.4MB
            backup reference list: 20211006-000004F, 20211006-000004F_20211006-002005I

        incr backup: 20211006-000004F_20211006-030327I
            timestamp start/stop: 2021-10-06 03:03:27 / 2021-10-06 03:06:29
            wal start/stop: 00000085000000E10000003C / 00000085000000E10000003C
            database size: 10.2GB, database backup size: 1.1GB
            repo1: backup set size: 1.8GB, backup size: 84.5MB
            backup reference list: 20211006-000004F, 20211006-000004F_20211006-002005I, 20211006-000004F_20211006-025605I

        diff backup: 20211006-000004F_20211006-030913D
            timestamp start/stop: 2021-10-06 03:09:13 / 2021-10-06 03:14:14
            wal start/stop: 00000085000000E10000003E / 00000085000000E10000003E
            database size: 10.2GB, database backup size: 1.8GB
            repo1: backup set size: 1.8GB, backup size: 172.3MB
            backup reference list: 20211006-000004F

        incr backup: 20211006-000004F_20211006-032006I
            timestamp start/stop: 2021-10-06 03:20:06 / 2021-10-06 03:24:49
            wal start/stop: 00000085000000E100000040 / 00000085000000E100000040
            database size: 10.2GB, database backup size: 1.2GB
            repo1: backup set size: 1.8GB, backup size: 97.6MB
            backup reference list: 20211006-000004F, 20211006-000004F_20211006-030913D

        full backup: 20211006-033114F
            timestamp start/stop: 2021-10-06 03:31:14 / 2021-10-06 03:39:08
            wal start/stop: 00000085000000E100000042 / 00000085000000E100000042
            database size: 10.2GB, database backup size: 10.2GB
            repo1: backup set size: 1.8GB, backup size: 1.8GB

        incr backup: 20211006-033114F_20211006-034006I
            timestamp start/stop: 2021-10-06 03:40:06 / 2021-10-06 03:44:51
            wal start/stop: 00000085000000E100000044 / 00000085000000E100000044
            database size: 10.2GB, database backup size: 1.2GB
            repo1: backup set size: 1.8GB, backup size: 97.5MB
            backup reference list: 20211006-033114F

        incr backup: 20211006-033114F_20211006-040003I
            timestamp start/stop: 2021-10-06 04:00:03 / 2021-10-06 04:07:40
            wal start/stop: 00000085000000E100000046 / 00000085000000E100000054
            database size: 10.7GB, database backup size: 1.8GB
            repo1: backup set size: 1.8GB, backup size: 187.6MB
            backup reference list: 20211006-033114F
```

**Comment by dwsteele** [comment]
It would be helpful if you would give details about your environment as requested in the issue template.

Usually `unable to wait on child process: [10] No child processes` is caused by the OOM killer. It could be a segfault as well, but there's not enough information here to tell.

The stuck remote process issue is reproducible in 2.35 but not in head so I did a bisect this morning. It seems to have been fixed in 475b57c89b58e9, though I'm not sure why. It could have been a use-after-free issue since this commit delays freeing the object memory until the mem context is freed.

**Comment by dwsteele** [comment]
Try adding `--log-level-file=trace --log-subprocess` to your backup. if a local process fails get the log for that process, .e.g `/var/log/pgbackrest/<stanza>-backup-local-XXX.log`, where XXX is the local process that failed.  Post the last 1000 lines of that log and we'll have a look.

**Comment by todeb** [comment]
Thank you. It works with start-fast=y
Actually the checkpoint_timeout I had set to 30min. That is the time that wals are still avaiable for patroni replication.
I do not want to change that, so I think I just misuderstood how pgbackrest is working and I set very frequent backups (every 20min).
I set backups now to one per day. For rpo I will just need probably to adjust archive_timeout. Then do some testing of pitr and rto.

**Comment by dwsteele** [comment]
Glad that worked. You have checkpoint timeout set to the same value as our protocol timeout, so by the time the checkpoint happened the local worker processes had timed out. We are planning to add a warning in this case https://github.com/pgbackrest/pgbackrest/projects/2#card-41782593. I've bumped the priority on that since we've seen it again in the field.

Personally, I'm a fan of setting `archive_timeout=900`. That gives a reasonable RPO for most cases.
